tests/unit/test_user_traits.cc

```diff
@@ -0,0 +1,204 @@
+//
+// * Redistributions of source code must retain the above copyright notice,
+//   this list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact darma@sandia.gov
+//
+// *****************************************************************************
+//@HEADER
+*/
+
+#include "test_harness.h"
+
+#include <checkpoint/checkpoint.h>
+
+#include <gtest/gtest.h>
+
+#include <vector>
+#include <cstdio>
+
+namespace checkpoint { namespace tests { namespace unit {
+template <typename T>
+struct TestUserTraits : TestHarness { };
+TYPED_TEST_CASE_P(TestUserTraits);
+
+struct CheckpointTrait {};
+struct CheckpointTraitNonintrusive {};
+namespace CheckpointNamespace {
+struct CheckpointTraitNamespaced {};
+}
+
+struct ShallowTrait {};
+struct TraitPairA {};
+struct TraitPairB {};
+
+static constexpr int const u_val = 'a';
+struct UserObjectA {
+  UserObjectA() = default;
+  explicit UserObjectA(int val)
+    : name('A'), val_a(val) {};
+  explicit UserObjectA(int val, char name_)
+    : name(name_), val_a(val) {};
+
+  template <typename S>
+  void serialize(S& s) {
+    std::cout << "A: serializing with type "
+      << abi::__cxa_demangle(typeid(s).name(), nullptr, nullptr, nullptr)
+      << std::endl;
+    EXPECT_FALSE((S::template has_traits<ShallowTrait>::value));
+
+    EXPECT_EQ(
+      (S::template has_traits<TraitPairA>::value),
+      (S::template has_traits<TraitPairB>::value)
+    );
+
+    if constexpr(S::template has_traits<CheckpointTrait>::value){
+      s | name;
+    }
+    s | val_a;
+  }
+
+  char name;
+  int val_a;
+};
+
+template<
+  typename S,
+  typename = typename S::template has_traits<CheckpointTraitNonintrusive>::type
```

Comment:
use `enable_if` here, eventually will be replaced by `requires`

Reply:
also do we test if this gets called?

Reply:
We do check if this is called in, test_serialize_extra. It confirms that UserObjectA does not serialize the name without a given trait, and does when that trait is included. And here we test with the three traits:
```
REGISTER_TYPED_TEST_CASE_P(TestUserTraits, test_serialize_extra);    
using TraitTypes = ::testing::Types<CheckpointTrait, CheckpointTraitNonintrusive, CheckpointNamespace::CheckpointTraitNamespaced>;    
INSTANTIATE_TYPED_TEST_CASE_P(checkpoint_trait, TestUserTraits, TraitTypes, );
```

Technically we aren't verifying that the specific function is called based on the trait, but that would require our has_traits to be broken and I think we test that well enough with the checks for ShallowTrait removal and TraitPairA/B matching

Reply:
I added an extra check to ensure that our trait detection doesn't mix up two different traits, so as long as that passes and this passes we can be sure that we call the extra functionality and that the right one is called